### PR TITLE
chore: client changes to support separate servers

### PR DESCRIPTION
**Pull request by yhl25**
- introduced separate sdk clients for each service `mapper`, `mapstreamer`, `reducer`, `sourcetransformer` and `sinker`
- Each service now has its own applier interface.
- The map forwarder now accepts the `MapApplier` and `MapStreamApplier` interfaces
- The source forwarder will accept the `SourceTransformApplier` interface
- The datum types have been changed to their respective request and response types for each service based on recent proto changes.
- introduced `GRPCBasedMap`, `GRPCBasedMapStream` and `GRPCBasedReduce` structs which implement their corresponding applier interfaces.
- The unnecessary `clienttest` structs have been removed and `NewFromClient(client)` has been introduced for each SDK client, which can be used for unit tests.
- Some refactoring has been done to support recent SDK changes.

All the e2e tests are fixed except for a few Python tests, I have added TODOs for them. @kohlisid please fix them once you have python sdk changes ready.

### CODE REVIEW

File: pkg/sdkclient/mapper/client.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	c := new(client)
+	var conn *grpc.ClientConn
+	var sockAddr string
+	// Make a TCP connection client for multiprocessing grpc server
+	if serverInfo.Protocol == info.TCP {
+		// Populate connection variables for client connection
+		// based on multiprocessing enabled/disabled
+		if err := resolver2.RegMultiProcResolver(serverInfo); err != nil {
+			return nil, fmt.Errorf("failed to start Multiproc Client: %w", err)
+		}
```
Comment (vigith) [inline pkg/sdkclient/mapper/client.go:82]: extract these to shared util func?

File: pkg/sdkclient/mapstreamer/client.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+)
+
+// client contains the grpc connection and the grpc client.
+type client struct {
+	conn    *grpc.ClientConn
+	grpcClt mapstreampb.MapStreamClient
+}
+
+// New creates a new client object.
+func New(inputOptions ...Option) (Client, error) {
```
Comment (vigith) [inline pkg/sdkclient/mapstreamer/client.go:45]: extract

File: pkg/sdkclient/sourcetransformer/interface.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	"context"
+
+	transformpb "github.com/numaproj/numaflow-go/pkg/apis/proto/sourcetransform/v1"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+// Client contains methods to call a gRPC client.
+type Client interface {
+	CloseConn(ctx context.Context) error
+	IsReady(ctx context.Context, in *emptypb.Empty) (bool, error)
```
Comment (vigith) [inline pkg/sdkclient/sourcetransformer/interface.go:29]: is this worth extracting out?

File: pkg/udf/rpc/errors.go
Comment (vigith) [inline pkg/udf/rpc/errors.go:1]: please add a doc.go

File: pkg/sdkclient/mapper/client.go (hunk 1 above)
Comment (yhl25) [reply pkg/sdkclient/mapper/client.go]: done

File: pkg/sdkclient/sourcetransformer/interface.go (hunk 3 above)
Comment (yhl25) [reply pkg/sdkclient/sourcetransformer/interface.go:29]: not in this PR, maybe a separate pr?

File: pkg/udf/rpc/errors.go
Comment (yhl25) [reply pkg/udf/rpc/errors.go:1]: done

**Review by vigith — APPROVED** [review]

File: pkg/sdkclient/sourcetransformer/interface.go (hunk 3 above)
Comment (vigith) [reply pkg/sdkclient/sourcetransformer/interface.go:29]: sure

**Pull request merged**